pkg/sentry/platform/ring0/pagetables/pagetables.go

```diff
@@ -41,13 +52,22 @@ type PageTables struct {
 	archPageTables
 }
 
-// New returns new PageTables.
-func New(a Allocator) *PageTables {
+// NewWithUpper returns new PageTables.
```

Comment:
As discussion elsewhere, you should document the API semantics of upperHalfPageTables here, and include any preconditions. E.g.

// NewWithUpper returns new PageTables.
//
// upperHalfPageTables are used for mapping the upper half of addresses, starting at upperLimit.
// These pageTables should not be touched (as invalidations may be incorrect) after they are
// passed as an upperHalfPageTables. Only when all dependent PageTables are gone may they
// be used. The intenteded use case is for kernel page tables, which are static and fixed.
//
// Precondition: upperLimit must be between canonical ranges.
// Precondition: ...

Reply:
Thinking about it a bit further, it does make me nervous that someone could accidentally do something with kernelPageTables, and it would be *really* hard to identify the source of that bug. (Since e.g. page table memory might just get reused.)

I like the simplicity of just checking for things in Map/Unmap/etc. Can we just do something like this?

```go
// MakeShared marks these PageTables as shared.
func (p *PageTables) MakeShared() {
   p.allocator.Drain() // Won't be modified.
   p.shared = true
}

// ...
// Precondition: p must not be shared.
func (p *PageTables) Map(...) {
   if p.shared {
       panic(...)
   }
   ...
}

// ...
// Precondition: upperHalfPageTables must be shared.
func NewWithUpper(a Allocator, upperHalfPageTables *pageTables) *PageTables {
    if upperHalfPageTables != nil {
         if !upperHalfPageTables.shared {
              panic(...)
         }
    }
    ...
}
```

That way everything is pretty simple (like the design you have here), and also still safe.

Reply:
One more good idea here: you can actually add an arch-hook for makeShared. E.g.

```go
// MakeShared marks these PageTables as shared.
func (p *PageTables) MakeShared() {
   p.shared = true
   w := sharedWalker{
      pageTables: p,
      visitor: sharedVisitor{}
   }
   w.iterateRange(0, ^0)
   p.allocator.Drain() // Won't be modified.
}
```

On amd64, this could go and mark all relevant PTEs as G, so they are shared across all address spaces.

Reply:
I like the idea to add the shared field in the Pagetables to make it read only. But I don't think we need to add a new sharedVisitor, we can just use G bit in mapUpperHalf() (and add X86_CR4_PGE bit in CR4).